Added article for testing. Added article to navigation. (#915)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/test-mule-applications.adoc

= Test Mule Applications

> ifndef::env-site,env-github[]

Reviewer: Hi Cristian,
So, I edited some very similar text for Fer yesterday. You might want to compare notes. His PR is https://github.com/mulesoft/docs-munit/pull/86/files

include::_attributes.adoc[]
endif::[]

> Building a comprehensive suite of automated tests for a Mule Application enables you to detect any regression or incompatible change in your application during tests in a local environment. MUnit is a Mule testing framework that lets you easily automate testing Mule applications, using both _unit_ and _integration_ tests. MUnit also provides a set of tools, such as a processor mocking framework that lets you test units of code in isolation.

Reviewer: ```suggestion
Building a comprehensive suite of automated tests for a Mule application enables you to detect any regression or incompatible change in your application by testing in a local environment. MUnit is a Mule testing framework that lets you easily automate testing Mule applications, using both _unit_ and _integration_ tests. MUnit also provides a set of tools, such as a processor-mocking framework that lets you test units of code in isolation.
```

Reviewer: Note that this introduction prepares the audience for the subtexts on unit testing and integration testing, but does not prepare the audience for the other two subtopics dealing with writing testable code and writing tests themselves. Can you add something that does that?

See the xref:munit::index.adoc[MUnit] documentation for more details.

> You can also test your Mule application's performance by profiling Mule runtime engine (Mule). This process can help you identify memory or performance-related issues in Mule applications or custom Mule extensions. To profile Mule, you need to load a Java Profiler into your on-premises Mule instance.

Reviewer: This seems like a different topic from testing the application; this is more about performance than unit testing, integration testing, writing testable code, or writing tests.

See xref:profiling-mule.adoc[Profiling Mule] for more instructions.

== Unit Tests

> The concept of unit tests may vary according to the programming paradigm that you are working on, but it always points to the same core concept: _to validate the correctness of an individual unit of source code_.

Reviewer: ```suggestion
The concept of unit tests varies according to your programming paradigm, but its core concept is always _to validate the correctness of an individual unit of source code_.
```

> In the context of unit testing, a unit of code is the smallest testable part of an application. What exactly constitutes the smallest testable part of an application depends on the application. When testing Mule applications, MuleSoft considers that the smallest testable part is a Mule flow or a sub-flow.

Reviewer: ```suggestion
In the context of unit testing, a unit of code is the smallest testable part of an application. What exactly constitutes the smallest testable part of an application depends on the application. For Mule applications, the smallest testable part is a Mule flow or a subflow.
```

Test cases must be independent of each other and it is the developer’s responsibility to ensure test inter-independency.
The code you test might interact with other units of code or components, always consider this possibility when designing unit tests. A good unit test ensures the isolation of the unit of code being tested, to avoid mistaking failure in other components for failure in the unit of code being tested.

> To isolate the target unit of code, use tools such as the `mock-when` processor provided by MUnit.

Reviewer: Consider combining and shortening to avoid repetition, something like:

As a developer, you must evaluate if the code you want to test interacts with other units of code or components and ensure that you design your unit test so that it isolates the unit of code being tested. Isolation ensures that any test failure relates to the code being tested, and not to some other code or component. To isolate your target code unit, use tools such as the `mock-when` processor provided by MUnit.

See xref:munit::mock-before-after-foreach-cookbook.adoc[Mocking a Message Before and Inside a Foreach Processor] for a usage example of the `mock-when` processor.

== Integration Tests

> Units of code collaborate between them to create an actual application. You test individual units of code with unit tests, and you test how units of code collaborate between them with integration tests.

Reviewer: ```suggestion
Because units of code collaborate to create an actual application, after you successfully test those units, you are ready to test your whole app through integration tests.
```

> The goal of an Integration test is _to validate that different units of code and modules work together as intended_.

Reviewer: ```suggestion
The goal of an integration test is _to validate that different units of code and modules work together as intended_.
```

> Integration tests must aggregate units of code modules that have already been successfully unit-tested. For this reason, always run your integration tests after your unit tests.

Reviewer: I suggest omitting this paragraph based on the suggested editorial addition to the intro sentence.

> Depending on the nature of your application, integration tests may require sandboxes, that is, environments for your application to connect to. When working with sandboxes, ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.

Reviewer: ```suggestion
Depending on the nature of your application, your integration tests might require a  _sandbox_ testing environment, which enables isolated testing evaluation outside of the working application environment. You must ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.
```

Reviewer: Please carefully evaluate the suggested rewording to ensure it does not change your intended meaning.
I wonder if we really need to define "sandbox"? Would this suffice:
```suggestion
Depending on the nature of your application, integration tests might require using a sandbox. If so, ensure that the state of the data in the sandbox is correct, both before and after running the integration test.
```
And doesn't this apply to any environment or test: that the state of the data be correct?

== Writing Testable Code

> The following guidelines help you create flows that are easier to test. Writing testable code implies creating a more configurable, extensible, and readable application.

Reviewer: ```suggestion
You can make your testing much easier (and create a more configurable, extensible, and readable application) by modularizing your code and defining execution environments.
```

* Modularize your code
+

> Breaking down your code in different files helps with readability. Do this by grouping flows that work together to achieve a common goal, or group code according to specific functional criteria.

Reviewer: ```suggestion
Breaking down your code into different files helps with readability. Do this by grouping flows that work together to achieve a common goal, or group code according to specific functional criteria.
```

* Write short flows
+

> Long flows are often hard to follow, harder to code, and harder to maintain. From the perspective of unit testing, long flows offer too many scenarios that can be triggered by a single point. Long flows might require you to perform very complex evaluations to validate a single scenario and are best avoided.

Reviewer: ```suggestion
Long flows can be hard to follow, hard to code, and hard to maintain, so it is best to avoid them. From the perspective of unit testing, long flows offer too many scenarios that can be triggered by a single point, and they require you to perform very complex evaluations to validate a single scenario.
```

* Define execution environments
+
Parameterize your code using placeholders. Normal use cases for this include addresses of outbound endpoints, such as DB or HTTP. Using placeholders enables you to modify the actual address when running tests (either unit or integration), and make it easier to promote your code between environments (DEV/QA/UAT/PROD).

> == Tips for Writing Tests

Reviewer: Hmm...I wonder if we can't make these recommendations more weighty than just "Tips"....
Test Writing Recommendations?
Best Practices for Writing Tests?
And...should this topic appear earlier in the set?

* Write your tests to be readable and maintainable.
* Name your test according to what you want to see when the test fails.

> * Use descriptive names for the error thrown by the test upon failure. This helps you avoid receiving wrongly-written or ambiguous error messages when the test fails.

Reviewer: List items and steps should not exceed one sentence or fragment. If necessary, start a new plain text para below the item.
```suggestion
* Use descriptive names for the error thrown by the test upon failure, to avoid receiving wrongly written or ambiguous error messages when the test fails.
```

> * Avoid long flows because they are harder to follow.

Reviewer: We covered this already in the code section; is this something different?

> * Code your test so that the scenario being covered and the failure conditions are explicit.

Reviewer: And doesn't this belong in the code writing section?

= Test Mule Applications
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Building a comprehensive suite of automated tests for a Mule Application enables you to detect any regression or incompatible change in your application during tests in a local environment.

Reviewer 2: This is more editorial, but in this first sentence I would a mention that you will cover unit test, integration test, writing testable code and its tips information, also about MUnit and Profiling Mule since these two definitions jump directly after this first sentence.
- Would these two product/features be considered to build automated tests? You can then add something like...Build automated test using MUnit or Profiling Mule...something like that.

= Test Mule Applications
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

Building a comprehensive suite of automated tests for a Mule Application enables you to detect any regression or incompatible change in your application during tests in a local environment.

> MUnit is a Mule testing framework that lets you easily automate testing Mule applications, using both _unit_ and _integration_ tests. MUnit also provides a set of tools, such as a message processor mocking framework that lets you test units of code in isolation.

Reviewer 3: message processor is a mule3 concept. now you should say just _processor_

See xref:profiling-mule.adoc[Profiling Mule] for more instructions.

== Unit Tests

The concept of unit tests may vary according to the programming paradigm that you are working on, but it always points to the same core concept: _to validate the correctness of an individual unit of source code_.

In the context of unit testing, a unit of code is the smallest testable part of an application. What exactly constitutes the smallest testable part of an application depends on the application. When testing Mule applications, MuleSoft considers that the smallest testable part is a Mule flow or a sub-flow.

Test cases must be independent of each other and it is the developer’s responsibility to ensure test inter-independency.
The code you test might interact with other units of code or components, always consider this possibility when designing unit tests. A good unit test ensures the isolation of the unit of code being tested, to avoid mistaking failure in other components for failure in the unit of code being tested.

> To isolate the target unit of code, use tools such as the `mock` message processor provided by MUnit.

Reviewer 3: `mock` -> `mock-when`
message processor -> processor

> See xref:munit::mock-before-after-foreach-cookbook.adoc[Mocking a Message Before and Inside a Foreach Processor] for a usage example of the `mock` message processor.

Reviewer 3: message processor -> processor

The goal of an Integration test is _to validate that different units of code and modules work together as intended_.
Integration tests must aggregate units of code modules that have already been successfully unit-tested. For this reason, always run your integration tests after your unit tests.

Depending on the nature of your application, integration tests may require sandboxes, that is, environments for your application to connect to. When working with sandboxes, ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.

== Writing Testable Code

The following guidelines help you create flows that are easier to test. Writing testable code implies creating a more configurable, extensible, and readable application.

* Modularize your code
+
Breaking down your code in different files helps with readability. Do this by grouping flows that work together to achieve a common goal, or group code according to specific functional criteria.

> Modularizing your code enables you to load only the portion of code that you want to test, avoiding the need to load unnecessary configurations during your test.

Reviewer 3: This is not provided by modularizing code (as it was in Mule3), in Mule 4 MUnit leverages Mule's lazy initialization functionality that only loads components used by the test
